File: lua-def-parser/src/lua_parser.rs
use std::{
    collections::HashSet,
    fmt::{self, Debug},
    str::FromStr,
};

use anyhow::Context;

use crate::parsing::StringExt as _;

pub fn parse_lua(input: &str) -> anyhow::Result<Vec<Result<LuaApiDef, String>>> {
    let mut lines = input.lines();

    let header = lines.next().context("Lua doc input file is empty")?;

    if let Some(version) = header.strip_prefix("Current modding API version: ") {
        tracing::info!(version, "Found lua doc header")
    } else {
        tracing::warn!("Lua doc header not found")
    }

    lines
        .next()
        .context("Lua doc input file is missing the splitter")?;

    Ok(lines.map(|s| s.parse()).collect())
}

#[derive(Debug)]
pub struct LuaApiDef {
    pub name: String,
    pub params: Vec<LuaApiParam>,
    pub ret: Option<LuaType>,
    pub doc: Option<String>,
}

impl FromStr for LuaApiDef {
    type Err = String;

    fn from_str(def: &str) -> Result<Self, Self::Err> {
        tracing::debug!(def, "Parsing lua definition");

        let (name, rest) = def.split_once('(').ok_or_else(|| def.to_owned())?;
        let (params, rest) = rest.split_once(')').ok_or_else(|| def.to_owned())?;

        let params = params.split(',').collect::<Vec<_>>();
        // typical split thing, eh
        let params = if params == [""] { Vec::new() } else { params };

        let params = params.into_iter().map(LuaApiParam::parse).collect();

        let (maybe_ret, maybe_doc) = rest
            .split_once("[")
            .or_else(|| rest.split_once('(')) // "Debugish function" ones
            .unwrap_or((rest, ""));

        let ret = maybe_ret.trim().strip_prefix("->").map(LuaType::parse);

        let doc = match maybe_doc.trim() {
            "" => None,
            doc => Some(doc.strip_suffix("]").unwrap_or(doc).replace("\\n", "\n")),
        };

        Ok(Self {
            name: name.trim().to_owned(),
            params,
            ret,
            doc,
        })
    }
}

#[derive(Debug)]
pub struct LuaApiParam {
    pub name: String,
    pub ty: LuaType,
    pub default: Option<String>,
}

impl LuaApiParam {
    fn parse(s: &str) -> Self {
        let s = s.trim();
        let (rest, default) = s.split_once('=').unwrap_or((s, ""));
        let (name, ty) = rest
            .split_once(':')
            .or_else(|| rest.split_once(' ').map(|(a, b)| (b, a)))
            .unwrap_or((rest, ""));

        Self {
            name: name.trim().to_owned(),
            ty: LuaType::parse(ty),
            default: match default.trim() {
                "" => None,
                d => Some(d.to_owned()),
            },
        }
    }
}

pub enum LuaType {
    Basic(String),
    Named(String, Box<LuaType>),
    Array(Box<LuaType>),
    Table(Box<LuaType>, Box<LuaType>),
    Disjoint(Vec<Box<LuaType>>),
    ReturnList(Vec<Box<LuaType>>),
}

impl LuaType {
    /// Walk the return type and collect all the basic types
    /// For example, `named:a|b,c,{d|c}` would return `{"a", "b", "c", "d"}`
    pub fn breakdown(&self) -> HashSet<&str> {
        fn recur<'s>(tpe: &'s LuaType, set: &mut HashSet<&'s str>) {
            use LuaType::*;
            match tpe {
                Basic(s) => {
                    set.insert(s);
                }
                Named(_, ty) | Array(ty) => recur(ty, set),
                Table(k, v) => {
                    recur(k, set);
                    recur(v, set);
                }
                ReturnList(parts) | Disjoint(parts) => {
                    for part in parts {
                        recur(part, set)
                    }
                }
            }
        }

        let mut set = HashSet::new();
        recur(self, &mut set);
        set
    }

    // lol
    fn merge(self) -> Self {
        use LuaType::*;
        match self {
            Array(inner) => Array(inner.merge().into()),
            Table(k, v) => Table(k.merge().into(), v.merge().into()),
            Disjoint(parts) => {
                let mut merged = Vec::new();
                for part in parts {
                    match *part {
                        Disjoint(inner_parts) => {
                            for inner in inner_parts {
                                merged.push(inner.merge().into());
                            }
                        }
                        part => merged.push(part.merge().into()),
                    }
                }
                Disjoint(merged)
            }
            ReturnList(parts) => {
                let mut merged = Vec::new();
                for part in parts {
                    match *part {
                        ReturnList(inner_parts) => {
                            for inner in inner_parts {
                                merged.push(inner.merge().into());
                            }
                        }
                        part => merged.push(part.merge().into()),
                    }
                }
                ReturnList(merged)
            }
            other => other,
        }
    }

    fn do_parse(s: &str) -> Self {
        let s = s.trim();

        if let Some(inner) = s.strip_parens_balanced('(', ')') {
            return Self::parse(inner);
        }

        if let Some(inner) = s.strip_parens_balanced('{', '}') {
            return match &*inner.split_balanced('-') {
                [k, v] => Self::Table(Self::parse(k).into(), Self::parse(v).into()),
                _ => Self::Array(Self::parse(inner).into()),
            };
        }

        // now we do splits from lowest to highest precedence

        // return list types are the lowest
        let multi_parts = s.split_balanced(',');
        if multi_parts.len() > 1 {
            let parsed_parts = multi_parts
                .into_iter()
                .map(|part| Self::parse(&part).into())
                .collect();
            return Self::ReturnList(parsed_parts);
        }

        // then named types, to allow for `named:a|b`
        let mut chars = s.chars();
        let mut has_colon = false;
        let name = chars
            .by_ref()
            .take_while(|ch| {
                if *ch == ':' {
                    has_colon = true;
                    return false;
                }
                *ch == '_' || ch.is_ascii_alphanumeric()
            })
            .collect::<String>();
        let name = name.trim();
        if !name.is_empty() && has_colon {
            let rest = chars.collect::<String>();
            return Self::Named(name.to_owned(), Box::new(Self::parse(&rest)));
        }

        // disjoint types are tightest
        let disjoint_parts = s.split_balanced('|');
        if disjoint_parts.len() > 1 {
            let parsed_parts = disjoint_parts
                .into_iter()
                .map(|part| Box::new(Self::parse(&part)))
                .collect();
            return Self::Disjoint(parsed_parts);
        }

        // trim some inconsistencies between the return and the doc
        Self::Basic(s.trim_end_matches([' ', '.', '-']).to_owned())
    }

    pub fn parse(s: &str) -> Self {
        Self::do_parse(s).merge()
    }
}

impl Debug for LuaType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use LuaType::*;

        fn recur(f: &mut fmt::Formatter<'_>, ty: &LuaType, indent: usize) -> fmt::Result {
            if f.alternate() {
                write!(f, "{: ^w$}", "", w = indent * 2)?;
            }
            match ty {
                Basic(s) => return write!(f, "(basic {s:?})"),
                Named(name, inner) => {
                    write!(f, "(named {name:?}")?;
                    if f.alternate() {
                        writeln!(f)?;
                    }
                    recur(f, inner, indent + 1)?;
                }
                Array(inner) => {
                    write!(f, "(array")?;
                    if f.alternate() {
                        writeln!(f)?;
                    }
                    recur(f, inner, indent + 1)?;
                }
                Table(k, v) => {
                    write!(f, "(table")?;
                    for part in [k, v] {
                        if f.alternate() {
                            writeln!(f)?;
                        }
                        recur(f, part, indent + 1)?;
                    }
                }
                Disjoint(parts) => {
                    write!(f, "(disj")?;
                    for part in parts {
                        if f.alternate() {
                            writeln!(f)?;
                        }
                        recur(f, part, indent + 1)?;
                    }
                }
                ReturnList(parts) => {
                    write!(f, "(return-list")?;
                    for part in parts {
                        if f.alternate() {
                            writeln!(f)?;
                        }
                        recur(f, part, indent + 1)?;
                    }
                }
            }
            write!(f, ")")
        }
        recur(f, self, 0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! check {
        ($expr:expr, @$snap:literal) => {
            insta::assert_debug_snapshot!(LuaType::parse($expr), @$snap);
        };
    }

    #[test]
    fn test_types() {
        crate::init_logging().unwrap();

        check!("component", @r#"(basic "component")"#);

        check!("entity_scale:number", @r#"
        (named "entity_scale"
          (basic "number"))
        "#);

        check!("{number}", @r#"
        (array
          (basic "number"))
        "#);

        check!("number,component", @r#"
        (return-list
          (basic "number")
          (basic "component"))
        "#);

        check!("number|component", @r#"
        (disj
          (basic "number")
          (basic "component"))
        "#);

        check!("{number,component}", @r#"
        (array
          (return-list
            (basic "number")
            (basic "component")))
        "#);

        check!("number|component,entity_scale:number", @r#"
        (return-list
          (disj
            (basic "number")
            (basic "component"))
          (named "entity_scale"
            (basic "number")))
        "#);

        check!("(number|component),entity_scale:number", @r#"
        (return-list
          (disj
            (basic "number")
            (basic "component"))
          (named "entity_scale"
            (basic "number")))
        "#);

        check!("number|(component,entity_scale:number)", @r#"
        (disj
          (basic "number")
          (return-list
            (basic "component")
            (named "entity_scale"
              (basic "number"))))
        "#);

        check!("{int|number|string}|nil", @r#"
        (disj
          (array
            (disj
              (basic "int")
              (basic "number")
              (basic "string")))
          (basic "nil"))
        "#);

        check!("(nil|script_return_type)|(nil,error_string)", @r#"
        (disj
          (basic "nil")
          (basic "script_return_type")
          (return-list
            (basic "nil")
            (basic "error_string")))
        "#);

        check!("named_union:number|string|nil", @r#"
        (named "named_union"
          (disj
            (basic "number")
            (basic "string")
            (basic "nil")))
        "#);

        check!("named_union:number|string|nil,second_return:bool|number", @r#"
        (return-list
          (named "named_union"
            (disj
              (basic "number")
              (basic "string")
              (basic "nil")))
          (named "second_return"
            (disj
              (basic "bool")
              (basic "number"))))
        "#);

        check!("{string-int}|named:{int-string|number}", @r#"
        (disj
          (table
            (basic "string")
            (basic "int"))
          (named "named"
            (table
              (basic "int")
              (disj
                (basic "string")
                (basic "number")))))
        "#);
    }

    #[test]
    fn snapshot_types() {
        crate::init_logging().unwrap();

        let input = include_str!("../lua_api_documentation.txt");

        let parsed = parse_lua(input).unwrap();

        let mut types = parsed
            .iter()
            .map(|d| d.as_ref().unwrap())
            .flat_map(|d| d.params.iter().map(|p| &p.ty).chain(d.ret.iter()))
            .flat_map(|t| t.breakdown().into_iter())
            .collect::<HashSet<_>>()
            .into_iter()
            .collect::<Vec<_>>();
        types.sort();

        insta::assert_debug_snapshot!(types, @r#"
        [
            "",
            "bool",
            "bool_is_new",
            "boolean",
            "component_id",
            "error_string",
            "float",
            "float x",
            "float y",
            "function",
            "int",
            "int_body_id",
            "item_entity_id",
            "multiple types",
            "multiple_types",
            "name",
            "new_text",
            "nil",
            "number",
            "obj",
            "physics_body_id",
            "replace_existing_cells",
            "script_return_type",
            "string",
            "uint",
            "uint32",
            "x",
            "y",
        ]
        "#);
    }
}

File: typegen/src/lua_parser.ts
export interface LuaApiDef {
  name: string;
  params: LuaApiParam[];
  ret?: LuaType;
  doc?: string;
}

export interface LuaApiParam {
  name: string;
  ty: LuaType;
  default?: string;
}

export type LuaType =
  | { kind: "basic"; value: string }
  | { kind: "named"; name: string; inner: LuaType }
  | { kind: "array"; inner: LuaType }
  | { kind: "table"; key: LuaType; value: LuaType }
  | { kind: "disjoint"; parts: LuaType[] }
  | { kind: "return-list"; parts: LuaType[] };

export function parseLua(input: string): Array<LuaApiDef | string> {
  const lines = input.split(/\r?\n/);
  if (!lines.length) {
    throw new Error("Lua doc input file is empty");
  }

  const header = lines[0];
  const [_, version] = header.match(/Current modding API version: (.*)/) || [];
  if (!version) {
    throw new Error("Lua doc input file is missing the version header");
  }

  if (!lines[1]) {
    throw new Error("Lua doc input file is missing the splitter");
  }

  return lines.slice(2).map((line) => {
    const def = parseLuaApiDef(line);
    return def ? def : line;
  });
}

function parseLuaApiDef(def: string): LuaApiDef | null {
  const idx1 = def.indexOf("(");
  if (idx1 === -1) {
    return null;
  }
  const idx2 = def.indexOf(")", idx1);
  if (idx2 === -1) {
    return null;
  }

  const name = def.slice(0, idx1).trim();
  const paramsRaw = def.slice(idx1 + 1, idx2);
  const params =
    paramsRaw === "" ? [] : paramsRaw.split(",").map(parseLuaApiParam);

  const rest = def.slice(idx2 + 1);
  let maybeRet = rest;
  let maybeDoc = "";
  const idxBracket = rest.indexOf("[");
  const idxParen = rest.indexOf("(");
  if (idxBracket !== -1) {
    maybeRet = rest.slice(0, idxBracket);
    maybeDoc = rest.slice(idxBracket + 1, rest.lastIndexOf("]"));
  } else if (idxParen !== -1) {
    maybeRet = rest.slice(0, idxParen);
    maybeDoc = rest.slice(idxParen + 1, rest.lastIndexOf(")"));
  }

  const ret = maybeRet.trim().startsWith("->")
    ? parseLuaType(maybeRet.trim().slice(2))
    : undefined;

  const doc = maybeDoc.trim() ? maybeDoc.replace(/\\n/g, "\n") : undefined;

  return { name, params, ret, doc };
}

function parseLuaApiParam(s: string): LuaApiParam {
  const [rest, defaultVal] = s.trim().split("=");
  let [name, ty] = ["", ""];
  const colonIdx = rest.indexOf(":");
  if (colonIdx !== -1) {
    name = rest.slice(0, colonIdx).trim();
    ty = rest.slice(colonIdx + 1).trim();
  } else {
    const spaceIdx = rest.indexOf(" ");
    if (spaceIdx !== -1) {
      ty = rest.slice(0, spaceIdx).trim();
      name = rest.slice(spaceIdx + 1).trim();
    } else {
      name = rest.trim();
      ty = "";
    }
  }
  return {
    name,
    ty: parseLuaType(ty),
    default: defaultVal ? defaultVal.trim() : undefined,
  };
}

function parseLuaType(s: string): LuaType {
  return mergeLuaType(doParseLuaType(s.trim()));
}

function doParseLuaType(s: string): LuaType {
  if (!s) {
    return { kind: "basic", value: "" };
  }

  const stripped = stripParensBalanced(s, "(", ")");
  if (stripped) {
    return parseLuaType(stripped);
  }
  const stripped2 = stripParensBalanced(s, "{", "}");
  if (stripped2) {
    const dashIdx = stripped2.indexOf("-");
    if (dashIdx !== -1) {
      const k = stripped2.slice(0, dashIdx);
      const v = stripped2.slice(dashIdx + 1);
      return { kind: "table", key: parseLuaType(k), value: parseLuaType(v) };
    }
    return { kind: "array", inner: parseLuaType(stripped2) };
  }

  const commaParts = splitBalanced(s, ",");
  if (commaParts.length > 1) {
    return { kind: "return-list", parts: commaParts.map(parseLuaType) };
  }

  const colonIdx = s.indexOf(":");
  if (colonIdx > 0 && /^[a-zA-Z0-9_]+$/.test(s.slice(0, colonIdx))) {
    const name = s.slice(0, colonIdx);
    const rest = s.slice(colonIdx + 1);
    return { kind: "named", name, inner: parseLuaType(rest) };
  }

  const pipeParts = splitBalanced(s, "|");
  if (pipeParts.length > 1) {
    return { kind: "disjoint", parts: pipeParts.map(parseLuaType) };
  }

  return { kind: "basic", value: s.replace(/[ .-]+$/, "") };
}

function mergeLuaType(t: LuaType): LuaType {
  if (t.kind === "array") {
    return { kind: "array", inner: mergeLuaType(t.inner) };
  }
  if (t.kind === "table") {
    return {
      kind: "table",
      key: mergeLuaType(t.key),
      value: mergeLuaType(t.value),
    };
  }
  if (t.kind === "disjoint") {
    const merged: LuaType[] = [];
    for (const part of t.parts) {
      if (part.kind === "disjoint") {
        merged.push(...part.parts.map(mergeLuaType));
      } else {
        merged.push(mergeLuaType(part));
      }
    }
    return { kind: "disjoint", parts: merged };
  }
  if (t.kind === "return-list") {
    const merged: LuaType[] = [];
    for (const part of t.parts) {
      if (part.kind === "return-list") {
        merged.push(...part.parts.map(mergeLuaType));
      } else {
        merged.push(mergeLuaType(part));
      }
    }
    return { kind: "return-list", parts: merged };
  }
  return t;
}

// Splits a string by a separator, ignoring separators inside (), {}, []
function splitBalanced(s: string, sep: string): string[] {
  const result: string[] = [];
  let depth = 0;
  let last = 0;
  for (let i = 0; i < s.length; ++i) {
    if ("({[".includes(s[i])) {
      ++depth;
    }
    if (")}]".includes(s[i])) {
      --depth;
    }
    if (s[i] === sep && depth === 0) {
      result.push(s.slice(last, i));
      last = i + 1;
    }
  }
  result.push(s.slice(last));
  return result;
}

function stripParensBalanced(
  str: string,
  open: string,
  close: string,
): string | undefined {
  if (!str.startsWith(open) || !str.endsWith(close)) {
    return undefined;
  }
  let depth = 0;
  for (let i = open.length; i < str.length - close.length; ) {
    const ch = str[i];
    if (ch === open) {
      depth++;
    } else if (ch === close) {
      if (depth === 0) return undefined;
      depth--;
    }
    i++;
  }
  return str.substring(open.length, str.length - close.length);
}
